### PR TITLE
Marketplace: Add the query parameters to product page URLs

### PR DESCRIPTION
**Pull request by kdevnel**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
In the old marketplace, product page URLs had certain query parameters. These are used for distinguishing merchants loading the product page from the in-app marketplace and showing a header to get them back on their store. 

On the new marketplace page, these query parameters are missing. This PR adds them back.

Closes https://github.com/Automattic/woocommerce.com/issues/17911

### How to test the changes in this Pull Request:

1. Checkout this branch
2. Run `pnpm run start` in the `packages/woocommerce-admin` folder
3. Run `composer dump-autoload` in the `packages/woocommerce` folder
4. Load [the discover page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions) (/wp-admin/admin.php?page=wc-admin&path=%2Fextensions) and click on a product card. You should end up at WooCommerce.com and you should see the in-app header.
5. Click on **Back to Store** and confirm you go back to your store.
    - That being said, the product page opens in a new tab. So not sure how useful is the **Back to Store** button.
7. Load [the browse page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions) (/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions) and click on a product card. You should end up at WooCommerce.com and you should see the in-app header.
8. Confirm the URL has `wccom-site`,`wccom-back`, `wccom-woo-version` and `wccom-connect-nonce` parameters
9. Click on **Back to Store** and confirm you go back to your store.
10. Click on a vendors name in the product card and confirm you end up WooCommerce.com again, with these set [as query parameters](https://github.com/woocommerce/woocommerce/pull/39901/files#diff-d33ce5ad0ffbbd53cd5beb965bbe34efd3f3dbf41a12e18c258dbd04eb02e6c2R26). 

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Fixes query parameters for Marketplace URLs

</details>

#### Screenshot
image

### CODE REVIEW

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx]: We'll be running this `getAdminSetting` for each `ProductCard` on the page. I wonder if it'll make things slower. Alternatively, we can move this logic to the `ProductListContextProvider` so that we'll run the `getAdminSetting` once and change all product URLs according to that. 

I'd be happy to make the change but wanted to see what do you think.

Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx]: Please go ahead! This was just the bare bones prototype to get things started.

Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx]: Ah, I found a problem in our approach with using the context. Unfortunately, since Discover doesn't yet utilise the context provider, we're unable to get this data from context on that page, making the issue still exist.

I do think it's the right approach eventually, but once we've moved Discover to using the context first. 

Until then, I suspect the better approach would be to take a similar, but less ideal approach and run the `getAdminSetting()` call in the `productList` component, so it's still only handled once. We can then pass the data into the cards from there.

Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx]: Alright, thanks for thinking this through! I added a bit of logic to the `ProductListContent` component. Same logic could also be implemented to the `ProductContent` though, I think not much difference there. 

That allowed me to keep the `ProductCard` as simple as possible. No more WCCOM logic in that component. It just prints whatever URL is passed to it.

**Comment by raicem** [comment]
Kyle opened the PR but I pushed a change that effectively moved a lot of things around. I can't add you as a reviewer as a result of it @kdevnel, but would be happy if you could take a look.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (kdevnel) [inline plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx]: I see we map here and then map again further down. Not critical, but I imagine if we improved the `appendURLParams()` method a bit, we could do this as part of the map method later in this file without too much messiness and save doing it twice.

That said, I don't think that's going to make any difference in performance.

**Review by kdevnel — COMMENTED** [review]
Single non-blocking comment. Works and looks good! I can't approve, since it's my PR so feel free to mark yourself as approved @raicem

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx]: Hmmm, yeah. That would probably make the code footprint much smaller and maybe more performant. 

I made that change and ended with these lines. I may be missing something because I don't fully get what could be improved with the `appendURLParams()` function that may result in a better looking code.

```jsx
<ProductCard
	key={ product.id }
	type="classic"
	product={ {
		title: product.title,
		icon: product.icon,
		vendorName: product.vendorName,
		vendorUrl: appendURLParams( product.vendorUrl, [
			[ 'utm_source', 'extensionsscreen' ],
			[ 'utm_medium', 'product' ],
			[ 'utm_campaign', 'wcaddons' ],
			[ 'utm_content', 'devpartner' ],
		] ),
		price: product.price,
		url: appendURLParams(
			product.url,
			Object.entries(
				wccomHelperSettings.inAppPurchaseURLParams
			)
		),
		description: product.description,
	} }
/>
```

Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx]: Looks good to me! I think this is the best approach over all.

Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx]: OK, I have pushed up this change and it's ready for a review.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx:17]: I'm not sure why we dropped `const { products } = props;` and added `props` here – I find the original method neat. But it's a debatable style question.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx:30]: We should've had this in the test steps! But I've checked it on both discovery and browse screens, and it reproduces what we have in WooCommerce 8.0.

I would've done the `appendURLParams` call separately, and just passed the result here, for better legibility. But that's another debatable stylistic choice.

**Review by andfinally — APPROVED** [review]
Looks good to me and checks out – sorry for the delay in reviewing!

**Comment by kdevnel** [comment]
Just noting for clarity that this is Cem's PR, I just opened it. I can't approve this PR as a result. I have approved it though so I'm okay with Cem "self-approving" the PR in my place if we need 2 approvals.

**Review by raicem — APPROVED** [review]

**Comment by raicem** [comment]
Thank you @kdevnel! Approved now. I'll merge this once I double check everything is good. I may have to do that next week though, I don't want to rush it since this will be merged into `trunk`.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx:17]: I think it was a stylistic decision. I personally find `	const { products } = props;` superfluous, especially when there is only one variable in props.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.tsx:30]: > We should've had this in the test steps! But I've checked it on both discovery and browse screens, and it reproduces what we have in WooCommerce 8.0.

I have added to the testing steps, although a bit late now. Still good to have it in there. 

> I would've done the appendURLParams call separately, and just passed the result here, for better legibility. But that's another debatable stylistic choice.

Totally understandable. Since this is in `Array.map`, I didn't want to turn that into a multiline code block and keep it a single statement.

**Pull request merged**
